### PR TITLE
Merge the query process of series index

### PR DESCRIPTION
**Pull request by ButterBright**
- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Fixes apache/skywalking#12239.

### CODE REVIEW

**Comment by wu-sheng** [comment]
Pleaae update e2e tool version to fix CI

File: banyand/internal/storage/index.go [hunk 1]
```diff
@@ -245,31 +223,49 @@ func (s *seriesIndex) Search(ctx context.Context, series []*pbv1.Series, opts In
...
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	var query index.Query
+	if opts.Query != nil {
+		query, err = s.store.Query(seriesMatchers, opts.Query)
+	} else {
+		query, err = s.store.Query(seriesMatchers, nil)
+	}
```
Comment (hanahmily) [inline banyand/internal/storage/index.go:238]:
```suggestion
	query, err = s.store.Query(seriesMatchers, opts.Query)
```

File: pkg/index/index.go [hunk 2]
```diff
@@ -156,7 +157,8 @@ type Writer interface {
 
 // FieldIterable allows building a FieldIterator.
 type FieldIterable interface {
-	Iterator(fieldKey FieldKey, termRange RangeOpts, order modelv1.Sort, preLoadSize int) (iter FieldIterator[*DocumentResult], err error)
+	Query(seriesMatchers []SeriesMatcher, secondaryQuery Query) (Query, error)
```
Comment (hanahmily) [inline pkg/index/index.go:160]: May we call it `BuildQuery`?

File: pkg/index/index.go [hunk 3]
```diff
@@ -204,8 +206,7 @@ type SeriesDocument struct {
 type SeriesStore interface {
 	Store
 	// Search returns a list of series that match the given matchers.
-	Search(context.Context, []SeriesMatcher, []FieldKey) ([]SeriesDocument, error)
-	Execute(context.Context, Query) (posting.List, error)
+	Search(context.Context, []SeriesMatcher, []FieldKey, Query) ([]SeriesDocument, error)
```
Comment (hanahmily) [inline pkg/index/index.go:209]: I might merge `SeriesMatcher` and `Query` into `Query`. The storage.index could build the query first, then pass it to the `Search`.

File: pkg/index/index.go [hunk 4]
```diff
@@ -156,7 +157,8 @@ type Writer interface {
 
 // FieldIterable allows building a FieldIterator.
 type FieldIterable interface {
-	Iterator(fieldKey FieldKey, termRange RangeOpts, order modelv1.Sort, preLoadSize int) (iter FieldIterator[*DocumentResult], err error)
+	Query(seriesMatchers []SeriesMatcher, secondaryQuery Query) (Query, error)
+	Iterator(fieldKey FieldKey, termRange RangeOpts, order modelv1.Sort, preLoadSize int, query Query, fields []string) (iter FieldIterator[*DocumentResult], err error)
```
Comment (hanahmily) [inline pkg/index/index.go:161]:
```suggestion
	Iterator(fieldKey FieldKey, termRange RangeOpts, order modelv1.Sort, preLoadSize int, query Query, fields []FieldKey) (iter FieldIterator[*DocumentResult], err error)
```

Please ensure that the semantics are consistent with `SeriesStore.Search`.

**Review by hanahmily — CHANGES_REQUESTED** [review]

File: pkg/index/inverted/inverted.go [hunk 5]
```diff
@@ -210,21 +210,31 @@ func (s *store) Iterator(fieldKey index.FieldKey, termRange index.RangeOpts,
...
 	}
+	query := bluge.NewBooleanQuery().AddMust(rangeQuery)
+	if indexQuery != nil && indexQuery.(*Query).query != nil {
+		query.AddMust(indexQuery.(*Query).query)
+	}
+	fields := make([]string, 0, len(fieldKeys))
+	for i := range fieldKeys {
+		fields = append(fields, fieldKeys[i].Marshal())
+	}
+	// TODO: add trace information
```
[the remark this answers is not in the document]
Comment (hanahmily) [reply pkg/index/inverted/inverted.go:234]: I agree. We should trace the query before merging it.

File: pkg/index/inverted/query.go [hunk 6]
```diff
@@ -49,30 +49,22 @@ type GlobalIndexError struct {
 
 func (g GlobalIndexError) Error() string { return g.IndexRule.String() }
 
+var _ index.Query = (*Query)(nil)
+
 // Query is a wrapper for bluge.Query.
 type Query struct {
```
Comment (hanahmily) [inline pkg/index/inverted/query.go:55]: Do you need to expose this type since its operations are wrapped by the `index.Query`?

**Review by hanahmily — CHANGES_REQUESTED** [review]

**Review by hanahmily — APPROVED** [review]
Nice job

**Comment by wu-sheng** [comment]
What is the new query performance benchmark after this?

**Comment by ButterBright** [comment]
Sorry for the late reply. I'm adding the relevant benchmarks. Theoretically, the topology queries should see a significant improvement.

**Pull request merged**